opensearch/catalog-opensearch-endpoint/src/main/java/org/codice/ddf/endpoints/ResourceUriTransformerCatalogFrameworkProxy.java

```diff
@@ -0,0 +1,199 @@
+        ResourceNotFoundException, ResourceNotSupportedException {
+        return catalog.getEnterpriseResource(request);
+    }
+
+    public Map<String, Set<String>> getEnterpriseResourceOptions(String metacardId)
+        throws ResourceNotFoundException {
+        return catalog.getEnterpriseResourceOptions(metacardId);
+    }
+
+    public ResourceResponse getLocalResource(ResourceRequest request) throws IOException,
+        ResourceNotFoundException, ResourceNotSupportedException {
+        return catalog.getLocalResource(request);
+    }
+
+    public Map<String, Set<String>> getLocalResourceOptions(String metacardId)
+        throws ResourceNotFoundException {
+        return catalog.getLocalResourceOptions(metacardId);
+    }
+
+    public ResourceResponse getResource(ResourceRequest request, String resourceSiteName)
+        throws IOException, ResourceNotFoundException, ResourceNotSupportedException {
+        return catalog.getResource(request, resourceSiteName);
+    }
+
+    public Map<String, Set<String>> getResourceOptions(String metacardId, String sourceId)
+        throws ResourceNotFoundException {
+        return catalog.getResourceOptions(metacardId, sourceId);
+    }
+
+    public Set<String> getSourceIds() {
+        return catalog.getSourceIds();
+    }
+
+    public SourceInfoResponse getSourceInfo(SourceInfoRequest sourceInfoRequest)
+        throws SourceUnavailableException {
+        return catalog.getSourceInfo(sourceInfoRequest);
+    }
+
+    public QueryResponse query(QueryRequest query) throws UnsupportedQueryException,
+        SourceUnavailableException, FederationException {
+        return catalog.query(query);
+    }
+
+    public QueryResponse query(QueryRequest queryRequest, FederationStrategy strategy)
+        throws SourceUnavailableException, UnsupportedQueryException, FederationException {
+        return catalog.query(queryRequest, strategy);
+    }
+
+    public BinaryContent transform(Metacard metacard, String transformerId,
+            Map<String, Serializable> requestProperties) throws CatalogTransformerException {
+        return catalog.transform(metacard, transformerId, requestProperties);
+    }
+
+    /**
+     * Converts the resource URIs found in the {@link SourceResponse} {@link Metacard}s using the
+     * {@link ActionProvider} injected in the constructor before delegating the call to the proxied
+     * {@link CatalogFramework} object.
+     */
+    public BinaryContent transform(SourceResponse response, String transformerId,
+            Map<String, Serializable> requestProperties) throws CatalogTransformerException {
+        transform(response);
+        return catalog.transform(response, transformerId, requestProperties);
+    }
+
+    public UpdateResponse update(UpdateRequest updateRequest) throws IngestException,
+        SourceUnavailableException {
+        return catalog.update(updateRequest);
+    }
+
+    private void transform(SourceResponse queryResponse) {
```

Comment:
I guess the question really boils down to the following: do we want this URI conversion to be (a) baked in specific endpoints or (b) do we want it to be optional/configurable for any endpoint?

If the answer is a) then in my mind the current solution using the proxy achieves this and also allows the logic to be re-used between endpoints that need the behavior to be "baked-in". Also, since it's internal, it doesn't require any analysis in terms of external and future impact.

If we feel that this conversion should be optional and we see a need for other pre-transform logic down the road, then I'm all for option b). However, if we go with that option, I feel that we should spend some time defining exactly what the plug-in interface should look like and what those plug-ins should be allowed to do. We may also want to consider writing a configurable plug-in that allows conversion of any attributes present in the response, not only the resource URI, and consider whether we want to keep using ActionProviders/Actions to do so.

In other words, I have nothing against the plug-in approach. I just think that before we expose a new plug-in and use it for this use case, we should spend some time thinking through all the use cases to make sure we expose the right thing and don't open the door to something we didn't intend.

Thoughts?

Eric